### PR TITLE
Issue 4 save favorites

### LINKED ISSUE #4: User can save a movie to their favorites list #4
## ⚙️ Feature

User can save a movie to their favorites list.

## 📐 Functionality

A user can save a movie to their favorites list by clicking on the heart located on each movie.

## 🖼️ Wireframes

Screenshot 2024-06-24 at 4 41 41 PM
Screenshot 2024-06-24 at 4 41 20 PM

## ✅ Task List

> ##### Commit checklist for tasks needed to complete the feature.

- [x] Save data to `localStorage`
  - [x] In `data.ts`:
    - [x] Create a `beforeunload` event listener:
      - [x] use `stringify` to convert the data object to JSON.
    - [x] Saved the JSON data to a variable using `getItem` to fetch previous JSON data.
        - [x] if there is no previous JSON data, then `stringify` jSON data and use `setItem` method to save to local storage.
     - [x] Create a conditional to check:
       - [x] if there's previous json data, then parse the previous saved JSON data.
- [x] In `data.ts`:
  - [x] Add a `favorites` property to the Movie interface of type `boolean`
  - [x] Add a `favorites` property to the Data interface of type `Movie[]`
- [ ] In `main.ts`:
  - [ ]  add font-awesome heart icons on `renderCard` function and place them on top of each card's movie poster.
  - [ ]  add heart icon to details view in `index.html`.
  - [ ] Create an `id` attribute for the icons to identify which one is clicked later.
- [ ] Add click event listener to heart and:
  - [ ] Use a conditional to check if the heart icon is already solid.
  - [ ] Change heart class using `classList` method to change heart to solid color.
  - [ ] Loop through movies array to find matching card based on id and push object to favorites list.
- **Clean Up**
  - [ ] Remove all buttons and placeholders that do not serve the functionality of the feature.
  - [ ] Remove all console logs.
  - [ ] Remove all commented out code.
  - [ ] Remove all CSS properties that have no effect.
  - [ ] Check all code for proper formatting and indentation.
  - [ ] Confirm that there are no errors in the console while using the application.
  - [ ] Confirm that all previous functionality still works without errors.
  - [ ] Confirm that the user interface looks natural on both mobile and desktop screens.
  - [ ] Create GIFs of the feature being used on both mobile and desktop screen sizes for inclusion in Pull Request.

### PR DESCRIPTION
**Pull request by techmatlock**
Recording 2024-06-26 at 10 16 45
Recording 2024-06-26 at 10 17 39

### CODE REVIEW

**Comment by techmatlock** [comment]
This PR includes a smooth sidebar transition, favorite/unfavorite, add data to local storage.

However, when you unfavorite right now it doesn't remove the object from local storage.  So you get duplicates if you unfavorite and then favorite again but I figure in the next feature (removing favorites) will cover that.

File: ts/data.ts [hunk 1]
```diff
@@ -13,17 +13,28 @@ interface Movie {
...
+const previousJsonData = localStorage.getItem('json-key-localstorage');
+
+if (previousJsonData !== null) {
+  data = JSON.parse(previousJsonData);
+}
+
+window.addEventListener('beforeunload', (): void => {
+  const jsonData = JSON.stringify(data);
+  localStorage.setItem('json-key-localstorage', jsonData);
+})
```
Comment (brandon-moy) [inline ts/data.ts:40]: Not a huge deal here but you would want to think about having a different key that is more specific to your application. You can do this later when you add a name to your project but something to think about.

**Comment by techmatlock** [comment]
Details view to favorite/unfavorite:

Recording 2024-06-26 at 10 27 11

**Comment by techmatlock** [comment]
Favorites view with favorited movies.  

Recording 2024-06-26 at 10 51 53

Also screenshot of the local storage with newly added favorited movies.
Screenshot 2024-06-26 at 10 45 19 AM

**Comment by techmatlock** [comment]
Local storage screenshot without any favorites

Screenshot 2024-06-26 at 10 54 20 AM

File: ts/main.ts [hunk 2]
```diff
@@ -251,9 +347,23 @@ $sidebarMenu.addEventListener('click', (event: Event): void => {
...
+  }
   viewSwap(selectedView);
-})
+});
+
+$detailsIcon.addEventListener('click', (): void => {
+  if ($detailsIcon.classList.contains('fa-solid')) {
+    $detailsIcon.className = 'fa-regular fa-heart fa-2xl details-icon';
+  } else if ($detailsIcon.classList.contains('fa-regular')) {
+    $detailsIcon.className = 'fa-solid fa-heart fa-2xl details-icon';
```
Comment (brandon-moy) [inline ts/main.ts:364]: Shouldn't this icon also add it to favorites when it makes it solid? (You don't need to have the functionality to remove from favorites but it should still add)

File: ts/main.ts [hunk 3]
```diff
@@ -217,30 +253,90 @@ function renderMovieDetails(data: Movie): void {
...
-      renderMovieDetails(moviesArr[i]);
+  const cardId = $card?.getAttribute('data-id');
+  if (!cardId) throw new Error('cardId not found.');
+
+  if ($eventTarget.matches('a')) {
+    for (let i = 0; i < moviesArr.length; i++) {
+      if (moviesArr[i].id === +cardId) {
+        viewSwap('learn-more');
+        console.log('moviesArr[i]', moviesArr[i]);
+        renderMovieDetails(moviesArr[i]);
```
Comment (brandon-moy) [inline ts/main.ts]: console.log

**Review by brandon-moy — CHANGES_REQUESTED** [review]
Looks good! There's only one main thing to think about with your details page heart icon. The other comments are a console.log and naming your local storage key (the local storage key is not a major issue to complete right away)

**Comment by techmatlock** [comment]
Added code to save to local storage in details view.

Recording 2024-06-26 at 11 19 43

**Review by brandon-moy — APPROVED** [review]
Approved!

**Pull request merged**